### DISCUSSION TITLE
error during generation of makefile

### DISCUSSION THREAD

**Discussion by brainbow1** [complexlogic/rsgain#44] [Q&A]
Hello!
I wanted to build form source, but during the generation of makefile, I get the following error:

`cmake .. -DCMAKE_BUILD_TYPE=Release
-- The CXX compiler identification is GNU 10.2.1
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.2") 
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE  
-- Checking for module 'libavcodec'
--   Found libavcodec, version 58.91.100
-- Checking for module 'libavformat'
--   Found libavformat, version 58.45.100
-- Checking for module 'libswresample'
--   Found libswresample, version 3.7.100
-- Checking for module 'libavutil'
--   Found libavutil, version 56.51.100
-- Checking for module 'taglib>=1.12'
--   Requested 'taglib >= 1.12' but version of TagLib is 1.11.1
CMake Error at /usr/share/cmake-3.18/Modules/FindPkgConfig.cmake:545 (message):
  A required package was not found
Call Stack (most recent call first):
  /usr/share/cmake-3.18/Modules/FindPkgConfig.cmake:733 (_pkg_check_modules_internal)
  CMakeLists.txt:70 (pkg_check_modules)

-- Configuring incomplete, errors occurred!
See also "/home/debian-desktop/rsgain/build/CMakeFiles/CMakeOutput.log".
See also "/home/debian-desktop/rsgain/build/CMakeFiles/CMakeError.log".
`

I also added the log files:
[CMakeError.log](https://github.com/complexlogic/rsgain/files/10185936/CMakeError.log)
[CMakeOutput.log](https://github.com/complexlogic/rsgain/files/10185937/CMakeOutput.log)

Please can u tell me how should I proceed?
I installed the respective dependencies before I started...

**Comment by complexlogic** [comment]
The reason is because of the minimum TagLib version requirement. I just relaxed it so you will be able to build with 1.11.1. Please run `git pull` and try again.

**Reply by phw** [reply]
@complexlogic Might be worth noting that many distributions, definitely Debian, have patched their 1.11 versions to fix the Ogg issue. So they are not affected by this.

**Reply by complexlogic** [reply]
@phw Thanks for the information. I believe the only major Linux distributions still using 1.11 are Debian stable and Ubuntu, all others have upgraded to 1.12 or later. In that case, the new warning may cause more confusion than benefit so it may be best to remove it.

**Reply by brainbow1** [reply]
thank you so much guys!
okay, I'm lookin forward to Debian bookworm stable anyway, so I think its only a matter of time till the update from 1.11 to 1.12.

The build/installation was now successfull, but I cant find the /usr/share/rsgain/presets anymore?

**Reply by brainbow1** [reply]
now I additionally installed the package that I have built, like its described in the github page.
I did:
`cmake .. -DPACKAGE=DEB

cmake .. -DCMAKE_INSTALL_PREFIX=/usr

sudo make package

sudo apt install ./rsgain_3.1.1_amd64.deb`

now the usr/share/rsgain/presets is available

but /usr/local/share/presets is available as well

What is the difference between those?
Did I mess anything up through my behaviour?
Seems like I have a double installation now...

**Reply by brainbow1** [reply]
I tried it with the following syntax and its stuck at 66.7%:

`debian-desktop@debian-desktop:~$ rsgain easy -m 10 -p mypreset /media/DATA/DATA/90s-test
[✔] Applying preset 'mypreset'...
[✔] Building directory tree...
[✔] Found 3 directories...
[✔] Scanning with 10 threads...
  66.7% Scanning directory '/media/DATEN/DATEN/90s-test/folder1'...
`

any idea what the problem could be?

**Reply by complexlogic** [reply]
When building the .deb package you need to specify the install prefix and the package variable in the same command. So your cmake command should be:
```bash
cmake .. -DCMAKE_INSTALL_PREFIX=/usr -DPACKAGE=DEB
```
In regard to your other issue: how many files do you have in those subdirectories? Is it a large number for each?

**Reply by brainbow1** [reply]
Okay, how can I clear all my mistakes and start from zero again?
Simply purging the installed rsgain package shouldnt be enough, because I built from source plus installed the package?
Sorry im not experienced in this area :(
The directory is a folder with two subfolders and overall file amount of 325 files, 10,1GB…

**Reply by complexlogic** [reply]
You can uninstall the files with `sudo make uninstall` from the directory where you built rsgain.

Note that in Easy Mode multithreaded, the progress update % is based on the *directories*, not on the files. Since you are scanning only two directories with many files, the progress will appear stuck, but the files are still being scanned. I am guessing that is the cause of your issue. Please be patient and wait for the scan to complete.

**Reply by brainbow1** [reply]
thanks a lot for ur help, I uninstalled both the package via apt, and the sourcebuild via sudo make uninstall, I am learning a lot during this process :-) And now I have just built from source without installing the deb package.

As Im quite unpatient I drastically reduced the file amount in the directories, but left both flac and mp3-files in the directories to test the main goal.
And it worked:

```
rsgain easy -m 10 -p mypreset /path/to/directory
[✔] Applying preset 'mypreset'...
[✔] Building directory tree...
[✔] Found 3 directories...
[✔] Scanning with 10 threads...

Scanning Complete
Time Elapsed:      40s
Files Scanned:     64
Clip Adjustments:  1 (1.6% of files)
Average Gain:      -6.44 dB
Average Peak:      1.005281 (0.05 dB)
Negative Gains:    61 (95.3% of files)
Positive Gains:    3 (4.7% of files)
```

What I don't understand though:
All replaygain-tags that are written to the files (flac and mp3) ar all lowercase.
I specified in the presetfile, that with MP3-files, the replaygaintags should be lowercase (lowercase=true), and with FLAC-files it should be uppercase (lowercase=false).
I did that because I have read in the documentation of "loudgain", that mp3-files need to have lowercase-tags in order to reach maximum flexibility/playability with most players... or did I get anything wrong here?
I think uppercase is the standard and most desirable , right?
As music player I use 
https://www.strawberrymusicplayer.org/
and
https://quodlibet.readthedocs.io/en/latest/
but most of the time strawberry...

**Reply by complexlogic** [reply]
> I specified in the presetfile, that with MP3-files, the replaygaintags should be lowercase (lowercase=true), and with FLAC-files it should be uppercase (lowercase=false).

Currently, any directory that has mixed file types uses the settings from the Global section. The MP3 section is only for when all files in the directory are MP3, same for FLAC, etc. It would require a lot of code change to allow the individual file sections apply to mixed directories, so I don't have time to do that now. 

> I did that because I have read in the documentation of "loudgain", that mp3-files need to have lowercase-tags in order to reach maximum flexibility/playability with most players... or did I get anything wrong here?

Yes, I am aware of the loudgain developer's recommendations for lowercase tags, however, I have found that not to be true in my experience. I also use Strawberry for my desktop music player and I can verify that it does read the standard uppercase tags. I recommend keeping the uppercase tags.

**Reply by brainbow1** [reply]
Ah okay, I see, so do I get it right, that with mixed directories its only possible to have "one specification for all", namely "Global", correct?
In this case, I would just delete the lowercase-stuff completely from the preset, and just live my life with uppercase-tags on all my files :-)
But does that also count for the option "ID3v2Version=4" , that I have set for MP3?
This version for is the "more up to date" version, right, and better to use with MP3?
If thats not the case then I could just leave the ID3v2Version as its specified by default and thats it?

**Reply by complexlogic** [reply]
I can add the Lowercase and ID3v2Version settings to Global, so they will apply to MP3 files.

**Reply by brainbow1** [reply]
alright but that will need lots of time doesnt it? No need to take any stress... 
As my knowledge doesnt reach that far I think its best when u decide if it makes sense to use ID3v2Version=4 for all mp3 files, if so, Im happy if u do so.
Regarding upper/lowercase the same, as long as everything works with uppercase-tags in Strawberry (with mp3 and flac), I think its enough to just have uppercase-tags globally, no matter if mp3 or flac...
but please let me know ur opinion/decision on this :)

**Reply by complexlogic** [reply]
Almost all modern players can handle both ID3v2.3 and ID3v2.4 so it doesn't make too much of a difference. I keep ID3v2.3 because it is *slightly* more compatible.

Anyway, I just pushed the change to allow Lowercase and ID3v2Version settings in Global section. You can `git pull` and rebuild if you want to use those settings in your preset.

**Reply by brainbow1** [reply]
Okay, thanks so much for ur help!
When I do not specify Lowercase in the preset, it will be globally uppercase, correct?

**Reply by brainbow1** [reply]
I tried with 
- no option set in the preset-file for Lowercase
- set "Lowercase=false" under Global
In both ways I get lowercase-replaygain-tags in my files as a result... what do I do wrong?
Everything else works fine, eraser-preset, scanning, all good, but all tags uppercase would be highly appreciated :-)

**Reply by complexlogic** [reply]
I'm unable to reproduce. Please double check your preset file.

**Comment by brainbow1** [comment]
two more questions:

for multi-threaded scanning:
I use "CPU: 11th Gen Intel i7-11700 (16) @ 4.800GHz ", and when I used rsgain with the "MAX"-option, my CPU seemed to work on the edge, didnt know if thats any good, especially for scanning a larger library later, so I reduced to 10 threads.. not sure if I can use the MAX without any risk?


in the options of rsgain it says:

` -s s, --tagmode=s         Scan files but don't write ReplayGain tags (default).
  -s d, --tagmode=d         Delete ReplayGain tags from files.
  -s i, --tagmode=i         Scan and write ReplayGain 2.0 tags to files.`

I decided for tagmode "i" because I want the tool to scan and write rg2.0 tags :-)
If I copy my preset, specifying tagmode "d", instead of "i", when I use this preset, all it does is deleting any ald replaygaintags that are written to the file, and nothing else, correct? So I could create a second preset just to use as an "eraser" ? :-)

**Reply by complexlogic** [reply]
> I use "CPU: 11th Gen Intel i7-11700 (16) @ 4.800GHz ", and when I used rsgain with the "MAX"-option, my CPU seemed to work on the edge, didnt know if thats any good, especially for scanning a larger library later, so I reduced to 10 threads.. not sure if I can use the MAX without any risk?

I'm not sure what you mean by "work on the edge", can you explain more?

> If I copy my preset, specifying tagmode "d", instead of "i", when I use this preset, all it does is deleting any ald replaygaintags that are written to the file, and nothing else, correct? So I could create a second preset just to use as an "eraser" ? :-)

Yes, you can make an eraser preset with `TagMode=d`. I also have one which I use for testing the program.

**Reply by brainbow1** [reply]
well I have a CPU-diagram in my Xfce taskbar, that visually shows me the occupancy rate of the CPU.
When I use the "MAX" option it goes into the red area, and my computer gets quite loud....
I have no experience with high performance stuff so I dont know if its easy for the cpu or if something is likely to overheat or so...
Have pretty much fans in my system though... maybe 14 threads would be an option to be not completely to the max but more efficient.

**Reply by complexlogic** [reply]
Ok, I understand. Scanning ReplayGain is a very intensive process, but it won't harm your system. You can do `-m MAX` safely. If it gets too hot, the OS will throttle.

**Comment by brainbow1** [comment]
How / which setting should be in the preset file? Should there stand anything with lowercase=true or lowercase=false or uppercase=true?Im unsure what to choose, right now there is lowercase=fault…What should there be to make it work? Is it set by default for upper or lowercase?

right now I use the following preset file:
`[Global]
TagMode=i
Album=false
TargetLoudness=-18
ClipMode=p
MaxPeakLevel=-1.0
TruePeak=true
Lowercase=false

[MP3]
#TagMode=i
#Album=false
#TargetLoudness=-18
#ClipMode=p
#MaxPeakLevel=-1.0
#TruePeak=true

[FLAC]
#TagMode=i
#Album=false
#TargetLoudness=-18
#ClipMode=p
#MaxPeakLevel=-1.0
#TruePeak=true
`

**Reply by complexlogic** [reply]
I just tested your preset and I got all upppercase tags. Which program are you using to verify the tags? I recommend VLC.

**Reply by brainbow1** [reply]
Ahhh okay, here was the problem:
I used "Ex Falso" to look up the replaygain-tags...I used it in the past for replaygaintags and writing tags like artist, title etc. manually in order to rename the files properly afterwards :) In future I will use Ex Falso only for the renaming stuff. For replaygain, now rsgain is my best friend, and I thank u for that :-)
when I use VLC instead of Ex Falso to display the meta data, it shows me the correct tags: all the way uppercase :-)

In the mid of 2023 I will install the new debian stable (bookworm), and during this process I think I can just take the latest .deb-package at that time, install it manually via apt and all should be fine, right?

Thank you so much for your help and patience with me.

I am so happy that I can use the tool now without any restrictions for my purpose. Thank you very much!

**Reply by complexlogic** [reply]
> In the mid of 2023 I will install the new debian stable (bookworm), and during this process I think I can just take the latest .deb-package at that time, install it manually via apt and all should be fine, right?

Yes
